stratum/hal/lib/barefoot/bf_pipeline_utils_test.cc

```diff
@@ -0,0 +1,240 @@
+            bfrt: "{\"bfrt\": 1}"
+            p4info {
+                actions {
+                    preamble {
+                        id: 3
+                        name: "my_pipe.drop"
+                        alias: "drop"
+                    }
+                }
+            }
+            pipelines {
+                name: "pipe"
+                scope: 0
+                scope: 1
+                scope: 2
+                scope: 3
+                context: "{\"ctx\":2}"
+                config: "<bin data>"
+            }
+        })pb";
+
+TEST(ExtractBfPipelineTest, FromProto) {
+  BfPipelineConfig bf_config;
+  ASSERT_OK(ParseProtoFromString(bf_config_1pipe_str, &bf_config));
+
+  std::string bf_config_bytes;
+  ASSERT_TRUE(bf_config.SerializeToString(&bf_config_bytes));
+
+  ::p4::v1::ForwardingPipelineConfig p4_config;
+  p4_config.set_p4_device_config(bf_config_bytes);
+
+  BfPipelineConfig extracted_bf_config;
+  EXPECT_OK(ExtractBfPipelineConfig(p4_config, &extracted_bf_config));
+  VLOG(1) << extracted_bf_config.DebugString();
+
+  EXPECT_TRUE(ProtoEqual(bf_config, extracted_bf_config));
+}
+
+TEST(ExtractBfPipelineTest, FromProto2Pipe) {
+  BfPipelineConfig bf_config;
+  ASSERT_OK(ParseProtoFromString(bf_config_2pipe_str, &bf_config));
+
+  std::string bf_config_bytes;
+  ASSERT_TRUE(bf_config.SerializeToString(&bf_config_bytes));
+
+  ::p4::v1::ForwardingPipelineConfig p4_config;
+  p4_config.set_p4_device_config(bf_config_bytes);
+
+  BfPipelineConfig extracted_bf_config;
+  EXPECT_OK(ExtractBfPipelineConfig(p4_config, &extracted_bf_config));
+  VLOG(1) << extracted_bf_config.DebugString();
+
+  EXPECT_TRUE(ProtoEqual(bf_config, extracted_bf_config));
+}
+
+TEST(ExtractBfPipelineTest, FromTarGzip) {
+  BfPipelineConfig bf_config;
+  ASSERT_OK(ParseProtoFromString(bf_config_tar_str, &bf_config));
+
+  // embedded my_pipe.tgz
+  //   Generated using:
+  //     tar --sort=name --owner=root:0 --group=root:0
+  //         --mtime='UTC 2019-01-01' -c . | gzip -9 | xxd -i
+  //   Contents of my_pipe.tgz:
+  //     ./
+  //     ./bfrt.json
+  //     ./my_prog.conf
+  //     ./p4info.txt
+  //     ./pipe/
+  //     ./pipe/context.json
+  //     ./pipe/tofino.bin
+  //   Contents of my_prog.conf:
+  //     {
+  //       "p4_devices": [{
+  //         "p4_programs": [{
+  //             "p4_pipelines": [{
+  //                 "context": "/tmp/pipe/context.json",
+  //                 "pipe_scope": [
+  //                     0,
+  //                     1,
+  //                     2,
+  //                     3
+  //                 ],
+  //                 "p4_pipeline_name": "pipe",
+  //                 "config": "/tmp/pipe/tofino.bin",
+  //                 "path": "/tmp"
+  //             }],
+  //             "bfrt-config": "/tmp/bfrt.json",
+  //             "program-name": "my_prog"
+  //         }],
+  //         "device-id": 1
+  //       }]
+  //     }
+  const unsigned char my_pipe_tgz_bytes[] = {
```

Comment:
The PacketIoManagerTest solves this quite well: https://github.com/stratum/stratum/blob/9f5bd2b285badbef11e81eca6c31d4a3c4342843/stratum/hal/lib/bcm/bcm_packetio_manager_test.cc#L1651

Reply:
fixed, what do you think?

Reply:
Looks much better. Good catch on the `sizeof()-1`, I'm pretty sure we got this wrong in a few places.